Fix #115: Add variants to Fluent documentation

flodolo opened the pull request.

stasm commented on tools/fluent/basic_syntax.md:
````diff
@@ -112,3 +112,27 @@ emails = { $unreadEmails ->
+sync-signedout-account-title = Connetti il tuo { -fxaccount-brand-name[lowercase] }
+```
+
+Similar to selectors, there must be a default variant, identified by `*`. Also notice that key names are arbitrary, and don’t need to be in English.
````
A technical nit: `selector` is what we call the expression before the `->` of a select expression. I think this should read: `Similar to select expressions, you must define a default variant…`.

flodolo commented:
Thanks, in my head "select expression" and "selector" where totally synonyms. I went through the rest of the doc and fixed a few places where I used them like that (and also used "strings").

Does the update cover all of your comments? Initially I added that extra variants in a message would be caught by compare-locales

stasm commented on tools/fluent/basic_syntax.md:
```diff
@@ -104,11 +104,39 @@ emails = { $unreadEmails ->
+
+## Variants and Terms
+
+As described at the beginning of the document, **terms** are a special type of **messages**: they can be recognized because of the identifier starting with a dash, e.g. `-brand-short-name`, and can only be referenced within other messages. More important, they are allowed to define additional attributes compared to the reference language (typically `en-US`).
```
I suggest mentioning the purpose terms serve. For example:

> As described at the beginning of the document, **terms** are a special type of **messages**. They are used to define translations of common words and phrases, which can then be used inside of other messages. They can be recognized by the identifier starting with a dash, e.g. `-brand-short-name`. Terms can also define additional language-specific attributes which are not present in the reference language (typically `en-US`).

stasm commented on tools/fluent/basic_syntax.md:
```diff
@@ -104,11 +104,39 @@ emails = { $unreadEmails ->
+
+As described at the beginning of the document, **terms** are a special type of **messages**: they can be recognized because of the identifier starting with a dash, e.g. `-brand-short-name`, and can only be referenced within other messages. More important, they are allowed to define additional attributes compared to the reference language (typically `en-US`).
+
+While in most cases terms will have a single value, it’s also possible to define multiple **variants**: different values, each one associated to a key, that can be selected when referencing the term ID from another message (or term). Unlike select expressions with a selector (e.g. the `emails` message in the previous section), the value retrieved doesn’t depend on external variables or functions.
```
Perhaps show the syntax here right away? I'd also remove the last sentence because it explains a pretty complex part of the design and I'm not sure it's worth distracting the reader with it.

> While in most cases terms will have a single value, it’s also possible to define multiple **variants**: different values, each one associated to a key. Variants represent different forms of the same value. They can be used to define grammatical cases or any other language-specific modifications of the value required by the grammar of the spelling rules. When referencing a term from another message, you can specify which variant to use with the `-term-identifier[variant name]` syntax.

> Consider the following example in English…

stasm commented on tools/fluent/basic_syntax.md:
```diff
@@ -104,11 +104,39 @@ emails = { $unreadEmails ->
+
+Similar to select expressions, you must define a default variant, identified by `*`. Also notice that key names are arbitrary, and don’t need to be in English.
+
+It’s important to note that this entire section applies to terms, not messages: while it’s possible to define variants in messages using a select expressions without a selector, such variants cannot be retrieved from other messages.
```
I'd remove this last sentence. Again, this is an obscure part of the design. I think the rest of the section makes it clear that variants only make sense when defined on terms.

flodolo commented:
Thanks, updated the doc.

stasm commented:
LGTM, thanks!
